src/libos/src/process/do_exec.rs

```diff
@@ -0,0 +1,86 @@
+use std::ffi::{CStr, CString};
+use std::path::Path;
+
+use super::do_exit::exit_old_process_for_execve;
+use super::do_spawn::new_process_for_exec;
+use super::process::ProcessFilter;
+use super::term_status::TermStatus;
+use super::wait::Waiter;
+use super::{do_exit, do_exit_group};
+use super::{table, ProcessRef, ProcessStatus};
+use super::{task, ThreadRef};
+use crate::interrupt::broadcast_interrupts;
+use crate::prelude::*;
+
+// FIXME: `occlum exec` command will return early if the application calls execve successfully.
+// Because the "execved"-ed application will run on a new thread and the current thread will exit.
+// `occlum run` will not have this problem.
+
+pub fn do_exec(
+    path: &str,
+    argv: &[CString],
+    envp: &[CString],
+    current_ref: &ThreadRef,
+) -> Result<isize> {
+    trace!(
+        "exec current process pid = {:?}",
+        current_ref.process().pid()
+    );
+
+    // 1. Freeze all child threads of current process to wait for exec status
+    let current = current!();
+    // 1.1 This must be locked to freeze other threads
+    let mut freeze = current.process().freeze();
+    *freeze = true;
+    // 1.2 Set process status to freeze
+    current.process().force_freeze(current_ref.tid());
+
+    // 2. Construct new process structure but with same parent, pid, tid
+    let new_process_ref = super::do_spawn::new_process_for_exec(path, argv, envp, current_ref);
+
+    // 3. Execute new main thread
+    if let Ok(new_process_ref) = new_process_ref {
+        let new_main_thread = new_process_ref
+            .main_thread()
+            .expect("the main thread is just created; it must exist");
+
+        // 4. Force exit all child threads of current process
+        // Once this succeeds, unfreeze and directly exit all threads.
+        let term_status = TermStatus::Exited(0 as u8);
+        current.process().force_exit(term_status);
+        // There might be some threads freezed at the sysret and wait for new status.
+        // Unfreeze these threads.
+        *freeze = false;
+        drop(freeze);
+        // Don't hesitate. Interrupt all threads right now (except the calling thread).
+        broadcast_interrupts();
+
+        // 5. Wait for all threads (except calling thread) to exit
+        let mut thread_num = current.process().threads().len();
+        while (thread_num != 1) {
+            thread_num = current.process().threads().len();
```

Comment:
The calling thread must be waiting here because it has to guarantee that the calling thread is the last exit thread to do some extra work. However, I didn't find out an easy way for the calling thread to wait for other threads to exit and not occupying the CPU. Added FIXME here.